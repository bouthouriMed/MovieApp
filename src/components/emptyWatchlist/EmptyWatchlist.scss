@import "@/styles//variables";

.empty-state {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
  padding: 3rem;
  color: $text-color;

  .empty-image {
    width: 200px;
    margin-bottom: 1.5rem;
    opacity: 0.8;
  }

  h2 {
    font-size: 1.5rem;
    margin-bottom: 0.5rem;
  }

  p {
    margin-bottom: 1.5rem;
    font-size: 1rem;
    color: $text-color;
  }

  .browse-btn {
    background-color: $primary-color;
    color: white;
    padding: 0.75rem 1.5rem;
    border-radius: 6px;
    font-weight: 500;
    text-decoration: none;
    transition: background 0.3s;

    &:hover {
      background-color: darken($primary-color, 10%);
    }
  }
}
