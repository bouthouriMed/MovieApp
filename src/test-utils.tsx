import React, { ReactNode } from "react";
import { render as rtlRender } from "@testing-library/react";
import { configureStore, Reducer } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import authReducer, { AuthState } from "./store/authSlice";
import watchListReducer, { WatchListState } from "./store/watchListSlice";
import { moviesApi } from "./store/movieSlice";

// --- Helper for renderHook with optional store ---
import {
  renderHook as rtlRenderHook,
  RenderHookOptions,
} from "@testing-library/react";
import { userApi } from "./store/userSlice";

// Explicit reducers type
export const reducers: {
  auth: Reducer<AuthState>;
  watchList: Reducer<WatchListState>;
  [moviesApi.reducerPath]: typeof moviesApi.reducer;
  [userApi.reducerPath]: typeof userApi.reducer;
} = {
  auth: authReducer,
  watchList: watchListReducer,
  [moviesApi.reducerPath]: moviesApi.reducer,
  [userApi.reducerPath]: userApi.reducer,
};

// Test helper
export const createTestStore = (preloadedState = {}) =>
  configureStore({
    reducer: reducers,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(moviesApi.middleware)
        .concat(userApi.middleware),
    preloadedState,
  });

// --- DOM mocks for scrollable elements ---
export const setupDOMMocks = () => {
  const originalOffsetWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "offsetWidth"
  );
  const originalScrollWidth = Object.getOwnPropertyDescriptor(
    HTMLElement.prototype,
    "scrollWidth"
  );

  beforeAll(() => {
    Object.defineProperty(HTMLElement.prototype, "offsetWidth", {
      configurable: true,
      value: 100,
    });
    Object.defineProperty(HTMLElement.prototype, "scrollWidth", {
      configurable: true,
      value: 500,
    });
  });

  afterAll(() => {
    if (originalOffsetWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "offsetWidth",
        originalOffsetWidth
      );
    if (originalScrollWidth)
      Object.defineProperty(
        HTMLElement.prototype,
        "scrollWidth",
        originalScrollWidth
      );
  });
};

// --- Unified render for components ---
interface RenderOptions {
  store?: ReturnType<typeof createTestStore>;
  wrapper?: React.ComponentType<{ children: ReactNode }>;
}

export const render = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (store) {
      return (
        <Provider store={store}>
          <MemoryRouter>{children}</MemoryRouter>
        </Provider>
      );
    }
    return <MemoryRouter>{children}</MemoryRouter>;
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};

export const renderHookWithStore = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  {
    store,
    wrapper,
    ...options
  }: RenderHookOptions<TProps> & {
    store?: ReturnType<typeof createTestStore>;
  } = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (wrapper) return React.createElement(wrapper, null, children);
    return store ? (
      <Provider store={store}>
        <MemoryRouter>{children}</MemoryRouter>
      </Provider>
    ) : (
      <MemoryRouter>{children}</MemoryRouter> // always wrap in Router
    );
  };
  return rtlRenderHook(hook, { wrapper: Wrapper, ...options });
};

export const renderWithoutRouter = (
  ui: React.ReactNode,
  { store, ...renderOptions }: RenderOptions = {}
) => {
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return store ? (
      <Provider store={store}>{children}</Provider>
    ) : (
      <>{children}</>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
};
