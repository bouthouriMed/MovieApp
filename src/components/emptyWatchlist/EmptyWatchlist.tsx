import { memo } from "react";
import { Link } from "react-router-dom";
import "./emptyWatchlist.scss";
import { ROUTES_URLS } from "@/routes";

function EmptyWatchlist() {
  return (
    <div className="empty-state">
      <p> Empty wishlist </p>
      <h2>Your wishlist is empty ðŸŽ¥</h2>
      <p>Start exploring movies and add them to your wishlist.</p>
      <Link to={ROUTES_URLS.Home} className="browse-btn">
        Browse Movies
      </Link>
    </div>
  );
}

export default memo(EmptyWatchlist);
