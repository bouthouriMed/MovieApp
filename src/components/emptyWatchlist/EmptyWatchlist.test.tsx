import { render } from "@/test-utils";
import EmptyWatchlist from "./EmptyWatchlist";

describe("EmptyWatchlist", () => {
  it("matches snapshot", () => {
    const { container } = render(<EmptyWatchlist />);

    expect(container).toMatchSnapshot();
  });
});
