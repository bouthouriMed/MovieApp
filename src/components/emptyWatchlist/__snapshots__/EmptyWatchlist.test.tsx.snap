// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmptyWatchlist > matches snapshot 1`] = `
<div>
  <div
    class="empty-state"
  >
    <p>
       Empty wishlist 
    </p>
    <h2>
      Your wishlist is empty ðŸŽ¥
    </h2>
    <p>
      Start exploring movies and add them to your wishlist.
    </p>
    <a
      class="browse-btn"
      data-discover="true"
      href="/"
    >
      Browse Movies
    </a>
  </div>
</div>
`;
